packages/loader/container-loader/src/container.ts

```diff
@@ -1045,29 +1045,38 @@ export class Container
 	}
 
 	public dispose(error?: ICriticalContainerError): void {
-		this._deltaManager.dispose(error);
-		this.verifyClosed();
+		this.trackCloseDisposeReentry(() => this._deltaManager.dispose(error));
 	}
 
 	public close(error?: ICriticalContainerError): void {
 		// 1. Ensure that close sequence is exactly the same no matter if it's initiated by host or by DeltaManager
 		// 2. We need to ensure that we deliver disconnect event to runtime properly. See connectionStateChanged
 		//    handler. We only deliver events if container fully loaded. Transitioning from "loading" ->
 		//    "closing" will lose that info (can also solve by tracking extra state).
-		this._deltaManager.close(error);
-		this.verifyClosed();
+		this.trackCloseDisposeReentry(() => this._deltaManager.close(error));
 	}
 
-	private verifyClosed(): void {
-		assert(
-			this.connectionState === ConnectionState.Disconnected,
-			0x0cf /* "disconnect event was not raised!" */,
-		);
+	private trackCloseDisposeReentryCalls = 0;
+	private trackCloseDisposeReentry(callback: () => void): void {
+		this.trackCloseDisposeReentryCalls++;
+		try {
+			callback();
+		} finally {
+			this.trackCloseDisposeReentryCalls--;
+		}
 
-		assert(
-			this._lifecycleState === "closed" || this._lifecycleState === "disposed",
-			0x314 /* Container properly closed */,
-		);
+		// We only want to verify connectionState and lifecycleState after close/dispose has fully finished
+		if (this.trackCloseDisposeReentryCalls === 0) {
```

Comment:
Alternate strategy could be to take advantage of the fact that these are synchronous, and chain the `verifyClosed` call via a dangling Promise.

Keeping track of incrementing/decrementing a gate always creeps me out, but in this case it is pretty straightforward.  Obviously embarking into the async world and using dangling promises has its own pitfalls...